nano/node/node.cpp

```diff
@@ -3737,6 +3742,86 @@ std::unique_ptr<seq_con_info_component> collect_seq_con_info (active_transaction
 	return composite;
 }
 }
+
+/**
+ * For all the blocks below this height which have been implicitly confirmed check if they
+ * are open/receive blocks, and if so follow the source blocks and iteratively repeat to genesis.
+ */
+void nano::node::add_confirmation_heights (nano::transaction const & transaction, nano::block_hash const & hash)
```

Comment:
When starting on a synced v18 ledger, this function appears to take a long time for some hashes right after startup. I noticed this because RPC was stuck. This call is in a write tx and all other IO threads ended up waiting for it to complete (specifically  `online_reps::sample`, `ongoing_store_flush`, `ongoing_peer_store` and thus RPC is blocked as well with 4 IO threads)

Example hash: F0394C8F7521DF8586BECAF2DA4868DA18E9D2DF702EAA89AFCABDB8AFF07EEB
This took ~30 minutes to complete, all other IO threads are eventually stuck on the write tx mutex.

Since `upgrade_v13_to_v14` sets height to 0, is the idea that add_confirmation_heights is part of a lazy upgrade, which may take a while for long chains the first time they are encountered?

One solution might be to process `open_receive_blocks` in time slices, like we do in `process_batch` - not sure if that fits with the stack processing logic.

Reply:
The initial ones being confirmed are expected to take a while. The issue we have is that for a confirmation height of an account to be accurate, all the receive blocks of the account also need to be confirmed, which if it hits a long account chain will take a very long time. I'm not sure if this can be broken up as any state in-between ends up invalidating it all completely. Perhaps there is a cleverer way this can be achieved, but how much of a problem is it for the initial upgrades? RPCs were down for people doing the sideband upgrades (days for some).

Reply:
>  but how much of a problem is it for the initial upgrades?

What confused me was that the node was operational for several minutes, until RPC stopped working.

It might be fine, I just wasn't aware of it. It may be a matter of documenting this effect on upgrade + clearing confirmation height.

Reply:
Ah right, I think I'm using term upgrade incorrectly here, as setting the confirmation height is really a continuous update. So the actual upgrade itself (v13_v14 adding 0 confirmation height) shouldn't take too long, at that point RPCs etc should work fine. It is only when blocks are confirmed off the network that the confirmation height is "updated", with the conditions mentioned earlier. I have added the documentation label to make sure this is not forgotten, thanks.